src/js/editor/editor.js

```diff
@@ -677,6 +695,45 @@ class Editor {
    * @public
    */
   run(callback) {
+    return this._run(callback);
+  }
+  /**
+   * This does the actual work of the public {run} above, actionType is used to group undos.
+   * Currently there are two action types:
+   * 1) `EDITOR_RUN_LAST_ACTION_STATE.insert_text` which is used to group any text insertion, and
+   * 2) `EDITOR_RUN_LAST_ACTION_STATE.delete` which contains any delete event.
+   *
+   * Any other event that happens such as insertion of cards, atoms, or paste events creates their own undo
+   * entries and are not grouped.
+   *
+   * @param {Function} callback Called with an instance of
+   *        {@link PostEditor} as its argument.
+   * @param {Number|undefined} actionType provides the type of the current
+   *                                      action, but only if it's a "runnable"
+   *                                      type.
+   * @return {Mixed} The return value of `callback`.
+   * @private
+   */
+  _run(callback, actionType) {
+
+    // Decides whether or not this current action is in a run of previous
+    // actions which dictates whether or not it overwrites the last item on the
+    // undo queue.
+    let isInARun = actionType && this._undoRunLastState === actionType;
+    this._undoRunLastState = actionType;
+
+    if(this._undoRunTimeout) {
+      clearTimeout(this._undoRunTimeout);
+      this._undoRunTimeout = null;
+    }
+
+    if(isInARun) {
+        this._undoRunTimeout = setTimeout(() => {
```

Comment:
This debouncing pattern isn't free- setting and clearing timeouts can be very slow JS code. Better might be to schedule a flush of this state in the future, and at the time of the flush check for if there have been any more recent actions that mean the flush should be cancelled. Then, if the flush is cancelled, schedule a new followup. This would be move the timeout setup/cancel work outside of the input and interactive parts of the flow.